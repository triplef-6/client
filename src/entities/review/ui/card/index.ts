export {Index as Card} from "./Index.tsx"