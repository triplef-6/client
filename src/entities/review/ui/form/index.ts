export {ReviewForm} from "./ReviewForm.tsx"
export {ReviewFormForUpdate} from "./ReviewFormForUpdate.tsx"