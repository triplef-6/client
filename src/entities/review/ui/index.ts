export {Index as Reviews} from "./Index.tsx"
export * from "./form"