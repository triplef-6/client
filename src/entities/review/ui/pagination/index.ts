export {Index as Pagination} from "./Index.tsx"