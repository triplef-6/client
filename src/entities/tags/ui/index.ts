export {Index as Tag} from "./Index.tsx"