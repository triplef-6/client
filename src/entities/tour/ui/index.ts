export {Tags} from "./tags"
export {Header} from "./header"
export {Carousel} from "./carousel"
export {Description} from "./description"
export {Sidebar} from "./sidebar"
export {Tours} from "./list"
export {TourCard} from "./card"