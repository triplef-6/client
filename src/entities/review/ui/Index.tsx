import {FC, useState} from "react";
import {Card} from "./card";
import {Header} from "./header";
import {Pagination} from "./pagination";
import {useReviewsByTourId} from "@/entities";
import {AppSkeleton} from "@/shared/ui";

type ReviewListProps = {
    tourId: number
    rating: number
    ratingCount: number
}

export const Index: FC<ReviewListProps> = ({rating, ratingCount, tourId}) => {

    const {data: reviews, isLoading, isError} = useReviewsByTourId(tourId)
    const [visibleReviews, setVisibleReviews] = useState<number>(2)

    if (isError) return null
    if (isLoading) return <AppSkeleton/>

    return (
        <div className={"flex flex-col gap-8 my-4"}>
            <Header
                rating={rating}
                ratingCount={ratingCount}
            />
            <div className={"flex flex-col gap-4"}>
                {reviews.slice(0, visibleReviews).map(review =>
                    <Card
                        key={review.id}
                        name={review.name}
                        rating={review.rating}
                        positiveText={review.positiveText}
                        negativeText={review.negativeText}
                        withChildren={review.withChildren}
                        personCount={review.personCount}
                    />
                )}
                <Pagination
                    visibleReviews={visibleReviews}
                    setVisibleReviews={setVisibleReviews}
                    reviews={reviews}
                />
            </div>
        </div>
    );
};