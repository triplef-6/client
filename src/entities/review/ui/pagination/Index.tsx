import {IReview} from "@/shared/types";
import React, {FC} from "react";
import {Button} from "@/shared/ui";

type PaginationProps = {
    visibleReviews: number
    setVisibleReviews:  React.Dispatch<React.SetStateAction<number>>
    reviews: IReview[]
}

export const Index: FC<PaginationProps> = ({visibleReviews, setVisibleReviews, reviews}) => {

    const loadTours = () => setVisibleReviews(prev => prev + 2)
    const unloadTours = () => setVisibleReviews(prev => prev - 2)

    return (
        <div className={"flex flex-row gap-4"}>
            {
                visibleReviews < reviews.length &&
                <Button onClick={loadTours} variant={"secondary"}>
                    Загрузить еще
                </Button>
            }
            {
                visibleReviews > 2 &&
                <Button onClick={unloadTours} variant={"secondary"}>
                    Скрыть
                </Button>
            }
        </div>
    );
};