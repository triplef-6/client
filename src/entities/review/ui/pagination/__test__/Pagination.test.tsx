import { describe, it, expect, vi } from "vitest"
import { render, screen } from "@testing-library/react"
import {Pagination} from "@/entities/review/ui/pagination"
import {userEvent} from "@testing-library/user-event";

describe("Pagination component", () => {

    const mockProps = {
        visibleReviews: 2,
        setVisibleReviews: vi.fn(),
        reviews: Array(5).fill({}),
    }

    it("Компонент должен рендериться", () => {
        render(<Pagination {...mockProps} />)
    })

    it("Snapshot тест", () => {
        const { container } = render(<Pagination {...mockProps} />)
        expect(container).toMatchSnapshot()
    })

    it("Кнопка 'Загрузить еще' должна отображаться, если есть дополнительные отзывы для загрузки", () => {
        render(<Pagination {...mockProps} />)
        expect(screen.getByText("Загрузить еще")).toBeInTheDocument()
    })

    it("Кнопка 'Скрыть' не должна отображаться, если visibleReviews равно 2", () => {
        render(<Pagination {...mockProps} />)
        expect(screen.queryByText("Скрыть")).not.toBeInTheDocument()
    })

    it("Функция setVisibleReviews должна быть вызвана с параметром +2, когда нажата кнопка 'Загрузить еще'", async () => {
        render(<Pagination {...mockProps} />)
        await userEvent.click(screen.getByText("Загрузить еще"))
        expect(mockProps.setVisibleReviews).toHaveBeenCalledWith(expect.any(Function))
    })

    it("Кнопка 'Скрыть' должна отображаться, если visibleReviews больше 2", () => {
        render(<Pagination {...mockProps} visibleReviews={4} />)
        expect(screen.getByText("Скрыть")).toBeInTheDocument()
    })

    it("Функция setVisibleReviews должна быть вызвана с параметром -2, когда нажата кнопка 'Скрыть'", async () => {
        render(<Pagination {...mockProps} visibleReviews={4} />)
        await userEvent.click(screen.getByText("Скрыть"))
        expect(mockProps.setVisibleReviews).toHaveBeenCalledWith(expect.any(Function))
    })

})