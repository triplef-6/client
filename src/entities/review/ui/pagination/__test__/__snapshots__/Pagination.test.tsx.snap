// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination component > Snapshot тест 1`] = `
<div>
  <div
    class="flex flex-row gap-4"
  >
    <button
      class="items-center gap-2 whitespace-nowrap text-base transition-colors focus-visible:outline-none disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-grayscale-300 text-grayscale-500 hover:opacity-90 flex justify-between h-12 px-10 rounded-xl"
    >
      Загрузить еще
    </button>
  </div>
</div>
`;
